add format and cleanup settings; reformat code accordingly

vladdu opened the pull request:
The settings are the default "Eclipse" built-in profiles, with the
following changes:

Format:
- use spaces to indent wrapped lines
- add missing new line at EOF
- reformat even comments
- allow on/off tags to suppress formatting

Clean up:
- organize imports
- remove trailing whitespace
- remove redundant type arguments

TL;DR: The default formatter has a page width of 120 chars, and 80
chars for comments. Indentation is with tabs.

vladdu commented:
Some of the ways the code is reformatted might not be to your liking :-) but IMHO it is better to have a consistent format.

mickaelistria commented:
The only one I'd argue here is "reformat even comments". Pretty often, for long lines for example, it's necassary to put it on several lines; when this happens, developers caring about readability use some purely human decision to know how to split: 1 line per argument, or 1 line per series of similar arguments, or for boolean to emphasize what the decision tree is, or if you think about a matrix, using rows and columns... For comments, it's the same. Breaklines, new lines or indentation can be important to emphasize ideas.
I don't think an automated formatter should in any case decide in place of a developer. Default rules are nice as suggestion when writing code, but if one decides it's better other way, it's to be respected by the formatter.

vladdu commented:
Ok, will remove those settings. 

There is a "Format Javadoc comments" that is enabled by default and does a pretty good job at doing the right thing for Javadoc (see the example in the formatter preview). Should that be enabled?

mickaelistria commented:
Yeah, I think it's fine for Javadoc since it allows HTML.

vladdu commented:
changed and reformatted everything

mickaelistria commented on org.eclipse.languageserver/src/org/eclipse/languageserver/LSPEclipseUtils.java:
```diff
@@ -42,8 +42,8 @@ public static int toOffset(Position position, IDocument document) throws BadLoca
-	public static TextDocumentPositionParamsImpl toTextDocumentPosistionParams(URI fileUri, int offset, IDocument document)
-			throws BadLocationException {
+	public static TextDocumentPositionParamsImpl toTextDocumentPosistionParams(URI fileUri, int offset,
+	        IDocument document) throws BadLocationException {
```
Do you like this kind of things? I'm not convinced about the benefit of limiting line width if it ends with many lines splitted.

mickaelistria commented on org.eclipse.languageserver/src/org/eclipse/languageserver/LSPEclipseUtils.java:
```diff
@@ -56,28 +56,31 @@ public static TextDocumentPositionParamsImpl toTextDocumentPosistionParams(URI f
 		IProject project = null;
 		for (IProject aProject : ResourcesPlugin.getWorkspace().getRoot().getProjects()) {
-			if (uri.startsWith(aProject.getLocationURI().toString()) && (project == null || project.getLocation().segmentCount() < aProject.getLocation().segmentCount())) {
+			if (uri.startsWith(aProject.getLocationURI().toString()) && (project == null
```
here it would be easier to read if the line break were before of after `&&` so we see the 2 operands.

mickaelistria commented on org.eclipse.languageserver/src/org/eclipse/languageserver/LSPStreamConnectionProviderRegistry.java:
```diff
@@ -67,15 +68,16 @@ private void initialize() {
 				IContentType contentType = contentTypeManager.getContentType(contentTypeId);
 				if (contentType != null) {
-					ILaunchConfiguration config = LaunchConfigurationStreamProvider.findLaunchConfiguration(launchType, launchName);
+					ILaunchConfiguration config = LaunchConfigurationStreamProvider.findLaunchConfiguration(launchType,
```
here too, I don't see how splitting the line makes it any better for maintainers.

mickaelistria commented on org.eclipse.languageserver/src/org/eclipse/languageserver/LSPStreamConnectionProviderRegistry.java:
```diff
@@ -102,15 +104,15 @@ private void persist() {
-			.filter(entry -> { return entry.getKey().equals(contentType); })
-			.map(entry -> { return new LaunchConfigurationStreamProvider(entry.getValue()); })
-			.toArray(StreamConnectionProvider[]::new));
+		return Arrays.asList(connections.stream().filter(entry -> {
```
the formatting here is much more complicated to read than earlier. It's a regular issue with automatic formatting when doing some "fluent APIs",

mickaelistria commented on org.eclipse.languageserver/src/org/eclipse/languageserver/LanguageServiceAccessor.java:
```diff
@@ -72,27 +73,28 @@ public boolean equals(Object obj) {
 
-	public static LanguageClientEndpoint getLanguageServer(IFile file, IDocument document, Predicate<ServerCapabilities> request) throws IOException {
+	public static LanguageClientEndpoint getLanguageServer(IFile file, IDocument document,
+	        Predicate<ServerCapabilities> request) throws IOException {
```
Is this really what we want?

mickaelistria commented on org.eclipse.languageserver/src/org/eclipse/languageserver/LanguageServiceAccessor.java:
```diff
@@ -72,27 +73,28 @@ public boolean equals(Object obj) {
 
-	private static ProjectSpecificLanguageServerWrapper getLSWrapper(IFile file, Predicate<ServerCapabilities> request) throws IOException {
+	private static ProjectSpecificLanguageServerWrapper getLSWrapper(IFile file, Predicate<ServerCapabilities> request)
+	        throws IOException {
```
Is this really what we want?

mickaelistria commented on org.eclipse.languageserver/src/org/eclipse/languageserver/LanguageServiceAccessor.java:
```diff
@@ -102,15 +104,14 @@ private static ProjectSpecificLanguageServerWrapper getLSWrapper(IFile file, Pre
-					|| wrapper.getServerCapabilities() == null /* null check is workaround for https://github.com/TypeFox/ls-api/issues/47 */
-					|| request.test(wrapper.getServerCapabilities())
-				)) {
+			if (wrapper != null && (request == null || wrapper
```
same here, formatting makes it more difficult to figure the operation tree.

mickaelistria commented on org.eclipse.languageserver/src/org/eclipse/languageserver/LanguageServiceAccessor.java:
```diff
@@ -120,13 +121,14 @@ private static ProjectSpecificLanguageServerWrapper getLSWrapper(IFile file, Pre
 			for (IContentType contentType : fileContentTypes) {
-				for (StreamConnectionProvider connection : LSPStreamConnectionProviderRegistry.getInstance().findProviderFor(contentType)) {
+				for (StreamConnectionProvider connection : LSPStreamConnectionProviderRegistry.getInstance()
+				        .findProviderFor(contentType)) {
```
Is this really what we want?

mickaelistria commented on org.eclipse.languageserver/src/org/eclipse/languageserver/LanguageServiceAccessor.java:
```diff
@@ -120,13 +121,14 @@ private static ProjectSpecificLanguageServerWrapper getLSWrapper(IFile file, Pre
-					if (request == null
-						|| wrapper.getServerCapabilities() == null /* null check is workaround for https://github.com/TypeFox/ls-api/issues/47 */
-						|| request.test(wrapper.getServerCapabilities())) {
+					if (request == null || wrapper
```
uglier, isn't it ;)

mickaelistria commented on org.eclipse.languageserver/src/org/eclipse/languageserver/languages/InitializeLaunchConfigurations.java:
```diff
@@ -42,7 +44,8 @@
 		ILaunchManager launchManager = DebugPlugin.getDefault().getLaunchManager();
-		ILaunchConfigurationType externalType = launchManager.getLaunchConfigurationType(IExternalToolConstants.ID_PROGRAM_LAUNCH_CONFIGURATION_TYPE);
+		ILaunchConfigurationType externalType = launchManager
+		        .getLaunchConfigurationType(IExternalToolConstants.ID_PROGRAM_LAUNCH_CONFIGURATION_TYPE);
```
Is this really better?

mickaelistria commented on org.eclipse.languageserver/src/org/eclipse/languageserver/languages/InitializeLaunchConfigurations.java:
```diff
@@ -62,13 +65,18 @@ public void earlyStartup() {
 				workingCopy.setAttribute(IExternalToolConstants.ATTR_LOCATION, "/usr/bin/node");
-				workingCopy.setAttribute(IExternalToolConstants.ATTR_TOOL_ARGUMENTS, "/home/mistria/git/omnisharp-node-client/languageserver/server.js");
+				workingCopy.setAttribute(IExternalToolConstants.ATTR_TOOL_ARGUMENTS,
+				        "/home/mistria/git/omnisharp-node-client/languageserver/server.js");
```
Is this really better?

mickaelistria commented on org.eclipse.languageserver/src/org/eclipse/languageserver/languages/csharp/CSharpKeywordsHighlighter.java:
```diff
@@ -32,121 +30,29 @@
-		"where",
-		"yield",
-	}));
+	private Set<String> keywords = new HashSet<>(Arrays.asList(new String[] { "abstract", "as", "base", "bool", "break",
```
This is exactly where an automatic formatter shows it doesn't really know how to communicate with humans ;)

mickaelistria reviewed (changes requested):
I've highlighted some cases where formatter makes code less readable and maintainable IMO.
However, I think the rules set up are good recommendations, but that some parts of the existing code already are better formatted than the formatter can do. So I would suggest to apply the .settings/ but leave the current code unchanged, and we'll adopt the formatting rules progressively as we update code.

vladdu commented:
Some of the ugly changes can be skipped by using @formatter:off and @formatter:on comments around the region in question. But it doesn't work for all cases.

I agree that some long lines with many operations will get split in weird places, but I would argue that this means that those operations should probably be extracted into functions, usually making the meaning of those operations more clear too. If you are open for this kind of changes, I can submit some for review, to make clear what I mean.

mickaelistria commented:
Let's keep item separated:
- Adding formatting rules as "recommendations" for new code: OK
- Making formatter applied automatically on existing code: Not, human often format for good reasons that formatter doesn't get
- Formatting all the existing code according to the rules: Not, as it reduces readability in multiple places
- Improving code by introducing meaningful variables or functions: Yes

vladdu commented:
All right, I will do so. Hopefully, after improving the code there will be few (if any) places affected negatively by the formatting.

vladdu committed:
add code format and cleanup settings

The settings are the default "Eclipse" built-in profiles, with the
following changes:

Format:
- use spaces to indent wrapped lines
- add missing new line at EOF
- allow on/off tags to suppress formatting

Clean up:
- organize imports
- remove trailing whitespace
- remove redundant type arguments

TL;DR: The default formatter has a page width of 120 chars, and 80
chars for comments. Indentation is with tabs.

vladdu commented:
Kept only the formatting options, to be used as the recommended settings.

mickaelistria commented:
Merged, thanks!
